solr/solrj/src/java/org/apache/solr/common/cloud/ZkStateReader.java

```diff
@@ -485,6 +506,20 @@ private void refreshLegacyClusterState(Watcher watcher)
           // Nothing to do, someone else updated same or newer.
           return;
         }
+        Set<String> liveNodes = this.liveNodes; // volatile read
+        for (Map.Entry<String, CollectionWatch> watchEntry : this.collectionWatches.entrySet()) {
+          String coll = watchEntry.getKey();
+          CollectionWatch collWatch = watchEntry.getValue();
+          ClusterState.CollectionRef ref = this.legacyCollectionStates.get(coll);
+          if (ref == null)
+            continue;
```

Comment:
Q: what happens if you try to set a watcher on a collection that doesn't exist yet?  Mostly curious.